src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -495,60 +492,61 @@ bool ShenandoahHeapRegion::oop_coalesce_and_fill(bool cancellable) {
       heap->old_generation()->card_scan()->coalesce_objects(obj_addr, fill_size);
       obj_addr = next_marked_obj;
     }
-    if (cancellable && ops_before_preempt_check-- == 0) {
-      if (heap->cancelled_gc()) {
-        suspend_coalesce_and_fill(obj_addr);
-        return false;
-      }
-      ops_before_preempt_check = preemption_stride;
+    if (cancellable && heap->cancelled_gc()) {
+      suspend_coalesce_and_fill(obj_addr);
+      return false;
     }
   }
   // Mark that this region has been coalesced and filled
   end_preemptible_coalesce_and_fill();
   return true;
 }
 
-// DO NOT CANCEL.  If this worker thread has accepted responsibility for scanning a particular range of addresses, it
-// must finish the work before it can be cancelled.
-void ShenandoahHeapRegion::oop_iterate_humongous_slice(OopIterateClosure* blk, bool dirty_only,
-                                                       HeapWord* start, size_t words, bool write_table) {
+size_t get_card_count(size_t words) {
   assert(words % CardTable::card_size_in_words() == 0, "Humongous iteration must span whole number of cards");
-  assert(is_humongous(), "only humongous region here");
-  ShenandoahGenerationalHeap* heap = ShenandoahGenerationalHeap::heap();
-
-  // Find head.
-  ShenandoahHeapRegion* r = humongous_start_region();
-  assert(r->is_humongous_start(), "need humongous head here");
   assert(CardTable::card_size_in_words() * (words / CardTable::card_size_in_words()) == words,
          "slice must be integral number of cards");
+  return words / CardTable::card_size_in_words();
+}
+
+void ShenandoahHeapRegion::oop_iterate_humongous_slice_dirty(OopIterateClosure* blk,
```

Comment:
We're removing the dirty_only bool argument?  Was that always true for all callers?  Or always true if write table?

Reply:
I see, I think.  You've introduced a new method to handle the not-dirty case.

Reply:
The original method was split it into two. The code path that handles _all_ cards didn't actually even need to look at the card table. The call sites where `dirty_only == false` call this new method. The call sites where `dirty_only == true` call this `oop_iterate_humongous_slice_dirty`.